modules/calib3d/test/test_cameracalibration.cpp

```diff
@@ -246,26 +246,59 @@ CV_ProjectPointsTest ProjectPoints_test;
 
 typedef Matx33d RotMat;
 
+/* Parameters to be modified during calibration */
+struct CalibratedParams
+{
+    std::vector<Vec3d>        transVects;
+    std::vector<RotMat>       rotMatrs;
+    std::vector<Point3d>      newObjPoints;
+    std::vector<double>       stdDevs;
+    std::vector<double>       perViewErrors;
+
+    Mat             cameraMatrix;
+    Mat             distortion = Mat::zeros(1, 5, CV_64F);
+};
+
+/* Parameters for input into calibration and verification of calibration.
+ * They are not modified during calibration. */
+struct FixtureParams
+{
+    std::vector<std::vector<Point2d> >  imagePoints;
+    std::vector<std::vector<Point3d> >  objectPoints;
+
+    std::vector<Vec3d>        goodTransVects;
+    std::vector<RotMat>       goodRotMatrs;
+    std::vector<Point3d>      goodObjPoints;
+    std::vector<double>       goodStdDevs;
+
+    Mat goodDistortion = Mat::zeros(1, 5, CV_64F);
+
+    double goodFcx,goodFcy;
+    double goodCx,goodCy;
+
+    Size imageSize;
+    Size etalonSize;
+    int numImages;
+    int numPoints;
+    int nstddev;
+    int flags;
+    int iFixedPoint;
+    bool releaseObject;
+};
+
 class CV_CameraCalibrationTest : public cvtest::BaseTest
 {
 public:
     CV_CameraCalibrationTest();
     ~CV_CameraCalibrationTest();
     void clear();
 protected:
-    int compare(double* val, double* refVal, int len,
+    int compare(const double* val, const double* refVal, int len,
                 double eps, const char* paramName);
-    virtual void calibrate(Size imageSize,
-        const std::vector<std::vector<Point2d> >& imagePoints,
-        const std::vector<std::vector<Point3d> >& objectPoints,
-        int iFixedPoint, Mat& distortionCoeffs, Mat& cameraMatrix, std::vector<Vec3d>& translationVectors,
-        std::vector<RotMat>& rotationMatrices, std::vector<Point3d>& newObjPoints,
-        std::vector<double>& stdDevs, std::vector<double>& perViewErrors,
-        int flags ) = 0;
-    virtual void project( const std::vector<Point3d>& objectPoints,
-        const RotMat& rotationMatrix, const Vec3d& translationVector,
-        const Mat& cameraMatrix, const Mat& distortion,
-        std::vector<Point2d>& imagePoints ) = 0;
+
```

Comment:
This class is refactored to make it easier to write custom tests with the data loaded in the `run` function.